isodatetime/datetimeoper.py

```diff
@@ -0,0 +1,341 @@
+                    dump_as_parsed=True)
+                parse_format = time_point.dump_format
+        if self.utc_mode:
+            time_point.set_time_zone_to_utc()
+        return time_point, parse_format
+
+    def date_shift(self, time_point, offset=None):
+        """Return a date string with an offset.
+
+        time_point -- A time point or time point string.
+                      Otherwise, use current time.
+
+        offset -- If specified, it should be a string containing the offset
+                  that has the format "[+/-]nU[nU...]" where "n" is an
+                  integer, and U is a unit matching a key in self.UNITS.
+
+        """
+        # Offset
+        if offset:
+            sign = "+"
+            if offset.startswith("-") or offset.startswith("+"):
+                sign = offset[0]
+                offset = offset[1:]
+            # Parse and apply.
+            try:
+                duration = self.duration_parser.parse(offset)
+            except ValueError:
+                raise OffsetValueError(offset)
+            if sign == "-":
+                time_point -= duration
+            else:
+                time_point += duration
+
+        return time_point
+
+    @staticmethod
+    def date_diff(time_point_1=None, time_point_2=None):
+        """Return (duration, is_negative) between two TimePoint objects.
+
+        duration -- is a Duration instance.
+        is_negative -- is "-" if time_point_2 is in the past of time_point_1.
+        """
+        if time_point_2 < time_point_1:
+            return (time_point_1 - time_point_2, "-")
+        else:
+            return (time_point_2 - time_point_1, "")
+
+    @classmethod
+    def date_diff_format(cls, print_format, duration, sign):
+        """Format a duration."""
+        if print_format:
+            delta_lookup = {
+                "y": duration.years,
+                "m": duration.months,
+                "d": duration.days,
+                "h": duration.hours,
+                "M": duration.minutes,
+                "s": duration.seconds,
+            }
+            expression = ""
+            for item in print_format:
+                if item not in delta_lookup:
+                    expression += item
+                elif float(delta_lookup[item]).is_integer():
+                    expression += str(int(delta_lookup[item]))
+                else:
+                    expression += str(delta_lookup[item])
+            return sign + expression
+        else:
+            return sign + str(duration)
+
+    @staticmethod
+    def get_calendar_mode():
+        """Get current calendar mode."""
+        return Calendar.default().mode
+
+    @classmethod
+    def set_calendar_mode(cls, calendar_mode=None):
+        """Set calendar mode for subsequent operations.
+
+        Raise KeyError if calendar_mode is invalid.
+
+        """
+        if not calendar_mode:
+            calendar_mode = os.getenv(cls.ENV_CALENDAR_MODE)
```

Comment:
This `if` looks a bit strange here. You set the calendar, but if no calender, than it gets a default value from environment variable. Couldn't we pass this responsibility to the caller?

e.g.

```diff
diff --git a/isodatetime/datetimeoper.py b/isodatetime/datetimeoper.py
index 3f445e6..aba8fc5 100644
--- a/isodatetime/datetimeoper.py
+++ b/isodatetime/datetimeoper.py
@@ -92,6 +92,8 @@ class DateTimeOperator(object):
         else:
             assumed_time_zone = None
 
+        if not calendar_mode:
+            calendar_mode = os.getenv(self.ENV_CALENDAR_MODE)
         self.set_calendar_mode(calendar_mode)
 
         self.time_point_dumper = TimePointDumper()
@@ -226,14 +228,12 @@ class DateTimeOperator(object):
         return Calendar.default().mode
 
     @classmethod
-    def set_calendar_mode(cls, calendar_mode=None):
+    def set_calendar_mode(cls, calendar_mode):
         """Set calendar mode for subsequent operations.
 
         Raise KeyError if calendar_mode is invalid.
 
         """
-        if not calendar_mode:
-            calendar_mode = os.getenv(cls.ENV_CALENDAR_MODE)
         Calendar.default().set_mode(calendar_mode)
 
     def strftime(self, time_point, print_format):
```

I think something similar is done with `ref_point_str`.